### DISCUSSION TITLE
Hantek6022API: Fix erased EEPROM

### DISCUSSION THREAD

**Discussion by ghost** [OpenHantek/OpenHantek6022#395] [Q&A] [answered]
**EDIT Ho-Ro:
This discussion belongs to [Hantek6022API](https://github.com/Ho-Ro/Hantek6022API) - it has no impact on OpenHantek6022.**

I tried program Scopevis.py for  plotting image, encountered error of (*attribute error and segmentation fault*) when the device is connected and reconnected  during program run.This fault was attempted by running reset_eeprom_6022.py.As it was working, the code was appended with scopevis.py,However when i runned the new code the epprom was reseted partially as result the Hanatek is not reponding to basic program. In order to make Hantek workable, it is required to reprogram it(resetting of its EEPROM bits). 
 usage: fx2upload [OPTIONS] FIRMWARE [VID PID]
  uploads firmware to a Cypress FX2 USB device identified by VID and PID
  OPTIONS:   -v verbose
  FIRMWARE:  intel hex file
  VID:       VendorID as hex value, default =  0x04B4
  PID:       ProductID as hex value, default = 0x8613
According to mentioned about quick hack procedure ,I tried but failed to reprogram.

Looking forward for your guidance.

**Comment by Ho-Ro** [comment]
> running reset_eeprom_6022.py

Who told you to do so?

>  the epprom was reseted partially

How do you know this?
How does the device announce itself on the bus? Please show the output of `lsusb`.
What does `dmesg` tell?

**Comment by ghost** [comment]
Thank you for quick reply :)
Here by sharing the output of lsub:
![Screenshot from 2024-02-13 10-28-15](https://github.com/OpenHantek/OpenHantek6022/assets/70463406/4c0a9e61-64f2-433c-b693-964f9a8f9468)
Here by sharing the output of dmesg:
![Screenshot from 2024-02-13 10-27-30](https://github.com/OpenHantek/OpenHantek6022/assets/70463406/d8155e34-4a69-48cd-ade2-f5f3cd7e61ba)

Even if it detecting on lsub and dmesg,,no response is shown here:
![Screenshot from 2024-02-13 10-44-46](https://github.com/OpenHantek/OpenHantek6022/assets/70463406/a2101f24-c24a-4000-aa14-0733a771166a)
   

Tried running this command but it doesn't workout after executing make file.  
 ./fx2upload.o -v /home/sapuser/Downloads/Hantek6022API/examples/dso6022be-firmware.hex 0x04B4 0x8613
bash: ./fx2upload.o: Permission denied

**Reply by Ho-Ro** [reply, marked as answer]
Ok, I see two issues
- the 1st is that you erased the EEPROM that contains the correct VID/PID of the Hantek - the shown default values of `0x04B4/0x8613` come directly from the FX2 chip when no valid EEPROM is seen. OH needs the Hantek values (`0x04B4/0x6022` for DSO6022-BE, or `0x04B4/0x602A` for the -BL variant).
- It also looks like you have no access rights for the USB port. You can use sudo to access the port. If this works OH6022 (as user) should work *temporarily* again until you unplug it. 
Then you can give the dangerous script `reset_eeprom_6022.py` another try, but make sure that the file `eeprom_6022.dat` is in the same directory.
Another posibility to fix the EEPROM content is to use my [fx2eeprom](https://github.com/Ho-Ro/fx2eeprom) tool:
```sh
./fx2eeprom w 0x04B4 0x8613 8 < eeprom_6022.dat
```

BTW, this is not an `OpenHantek6022` topic, but related to [Hantek6022API](https://github.com/Ho-Ro/Hantek6022API). But lets keep the discussion here. 
Nevertheless make sure that you have installed the `Hantek6022API` package correctly including the `udev` rules, `60-hantek6022api.rules` should go to `/etc/udev/rules.d`.
Please mind, this doesn't solve the rights for `0x04B/0x8613`, you still need the `sudo` to access the *"naked"* device.

**One last hint: Please make yourself familiar with the concept of the FX2 chip before fiddling with the EEPROM.**
Please understand the difference between **firmware upload** (this goes into the RAM of the FX2 and has to be done after each power on) and writing to the external EEPROM (the Hantek has a small EEPROM of 256 byte capacity, it can hold only the device ID and some data, in our case the calibration).
 
The [EZ-USB_TECHNICAL_REFERENCE_MANUAL](https://www.infineon.com/dgdl/Infineon-EZ-USB_TECHNICAL_REFERENCE_MANUAL-AdditionalTechnicalInformation-v08_00-EN.pdf?fileId=8ac78c8c7d0d8da4017d0f9093657d61) tells you everything you need to know, especially section 3.2 of TRM (EZ-USB Startup Modes).

One last question, you wrote 
> This fault was attempted by running reset_eeprom_6022.py.

Where did you get the info to use this tool, who told you to do so?
I ask because this is dangerous (as you have seen) and I want to understand your intention to shoot yourself in the foot.

**Comment by ghost** [comment]
Thank you for your valuable reply:
1. I had tried following with sudo command
    sudo  ./fx2upload.o -v /home/sapuser/Downloads/Hantek6022API/examples/dso6022be-firmware.hex 0x04B4 0x6022
    followed by reset_eeprom_6022.py (_it doesn't worked_)

2.I had tried another suggested method to fix
   using fx2eeprom tool.
  sudo  ./fx2eeprom w 0x04B4 0x8613 8 < eeprom_6022.dat

![Screenshot from 2024-02-14 17-46-05](https://github.com/OpenHantek/OpenHantek6022/assets/70463406/19129c61-15f5-4adc-947c-96e83f26307a)
(it doesn't work)
 
 Where did you get the info to use this tool, who told you to do so?
Answering to your question: While interfacing Hantek with Rpi I was interested to get graphical plot in the continuous loop.Instead of plotting multiple graphs, I wanted to refresh same graph so, I used the code of reset_eeprom_6022.py in while along with matplotlib function.As a result after execution of the program only partial bits of EEPROM was set. Because of which above issue arrised.

**Reply by ghost** [reply]
Thank you for your valuable reply:

    I had tried following with sudo command
    **sudo  ./fx2upload -v /home/sapuser/Downloads/Hantek6022API/examples/dso6022be-firmware.hex 0x04B4 0x8613
    followed by reset_eeprom_6022.py **(it  worked!!)****

  **Note: 1.fx2upload.o (removing .o extension) while executing command
             2.getting permission by right click on .o file.
             3.using dmsg finding out its VID and PID data 
             4.Executing above command.**
![Screenshot from 2024-02-14 18-53-46](https://github.com/OpenHantek/OpenHantek6022/assets/70463406/c48f1f06-23bf-4e4f-b19d-50b8d8ec3578)

The USB is detected!!
![Screenshot from 2024-02-14 18-55-19](https://github.com/OpenHantek/OpenHantek6022/assets/70463406/564545f6-96b3-4e3b-bf2e-747b18e967ca)

**Reply by Ho-Ro** [reply]
Is it detected also after unplugging and replugging?

**Reply by Ho-Ro** [reply]
Two more remarks:

>  I wanted to refresh same graph so, I used the code of reset_eeprom_6022.py in while along with matplotlib function.

Ok, you have a lessons learned:
**DO NOT USE `reset_eeprom_6022.py` UNLESS YOU KNOW WHAT YOU ARE DOING!**

> fx2upload.o (removing .o extension) while executing command

NEVER try to execute *.o files. These are *object* files that cannot be executed.
You link them to the startup and libraries to create the *executable* file (without any extension) - but that's just a convention with no technical necessity. You can call it whatever you want, but you create maximum confusion.

**Comment by ghost** [comment]
Yes it is detected after unplugging and replugging.

**Reply by Ho-Ro** [reply]
Perfect, you fixed it!

**Comment by ghost** [comment]
Thank you for the remark 😊
I will take care of the above mentioned points.
1.Will not use reset_eeprom_6022.py file,unless required.
2.Will never try to execute *.o files .

**Comment by ghost** [comment]
Hi @Ho-Ro  as earlier discussed, I'm working on the example program **scopevis.py.** I have two queries regarding the same.

1.**scopevis.py** runs properly for channel-1 but after configuring it for channel-2 according to the lib file **PyHT6022.LibUsbScope**.I am unable to plot a graph for channel 2.

2. Another example is **set_cal_out_freq_6022.py**  for calculating frequency, however, I'm unable to calculate the frequency of the waveform.

Looking forward to your help and guidance.

**Reply by Ho-Ro** [reply]
1. `scopevis.py` is old stuff w/o support.
2. please check latest commit in repo `https://github.com/Ho-Ro/Hantek6022API` that fixes `set_cal_out_freq_6022.py`.

**Reply by ghost** [reply]
Thank you for your valuable reply.

1. Okay, if not scopevis.py, is there any possible alternative for taking the signal from channel 2 of Hantek?

2. Okay, according to the latest commit repo, I get the following output.
usage: set_cal_out_freq_6022 FREQ
with 32 <= FREQ <= 100000
 
3.Is there any other method to calculate the numerical value of the frequency of the plotted wave by Hantek?

**Reply by Ho-Ro** [reply]
1: 
```usage: capture_6022.py [-h] [-d [DOWNSAMPLE]] [-g] [-o OUTFILE] [-r RATE] [-t TIME] [-x CH1] [-y CH2]

options:
  -h, --help            show this help message and exit
  -d [DOWNSAMPLE], --downsample [DOWNSAMPLE]
                        downsample 256 x DOWNSAMPLE
  -g, --german          use comma as decimal separator
  -o OUTFILE, --outfile OUTFILE
                        write the data into OUTFILE
  -r RATE, --rate RATE  sample rate in kS/s (20, 32, 50, 64, 100, 128, 200, default: 20)
  -t TIME, --time TIME  capture time in seconds (default: 1.0)
  -x CH1, --ch1 CH1     gain of channel 1 (1, 2, 5, 10, default: 1)
  -y CH2, --ch2 CH2     gain of channel 2 (1, 2, 5, 10, default: 1)
```
2: just select the output frequency you want, e.g. to create 3.4 kHz:
```
set_cal_out_freq_6022.py 3400
```
3: check the fft*.py examples

**Reply by ghost** [reply]
Okay, sir.Thank you.
